k8s: CEP GC runs in steps

raybejjani opened the pull request:
On large clusters the CEP GC would fetch the entire list of CEPs. These
can grow to significant sizes (100kB in one instance) and this causes
extreme degradation in kube-apiserver. We now iterate 10 CEPs at a
time, thus lowering the load we put on the cluster.

I also return the cache.Store object from our k8s helper factories. No functionality has changed there.

I began by trying to watch for CEPs in the controller but the code became a little complex. This was partly because only the node that should do GC needs to watch for CEPs, but that meant creating a non-shared informer (shared informers cannot stop once they begin watching for a specific type). In the end, limiting the list to a small number seemed like a reasonable compromise, especially since the GC isn't a critical operation here.

Relates to: https://github.com/cilium/cilium/issues/5913

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6029)
<!-- Reviewable:end -->

aanm committed:
k8s: Return cache.Store from k8s watches factory

The watcher also keeps a local cache of objects. We can use this for
lookups in places like garbage-collection.

Signed-off-by: Ray Bejjani <ray@covalent.io>

ianvernon commented on pkg/endpoint/endpoint.go:
```diff
@@ -197,36 +203,53 @@ func RunK8sCiliumEndpointSyncGC() {
+					ceps, err := ciliumClient.CiliumEndpoints(meta_v1.NamespaceAll).List(listOpts)
+					switch {
+					case ceps.Continue != "" && err != nil:
+						// this is ok, it means we saw a 410 ResourceExpired error but we
```
The comment is helpful, but is this documented somewhere? I'd suggest putting a link in a comment here to said documentation.

raybejjani replied:
Good point. I was, in fact, checking this incorrectly. I've fixed it.

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -197,36 +203,60 @@ func RunK8sCiliumEndpointSyncGC() {
-					clusterPodSet[podFullName] = true
-				}
+				var (
+					listOpts = meta_v1.ListOptions{Limit: 10}
```
Did you benchmark with 10/25/50?

raybejjani replied:
Not strictly, no. I did a run with 1k CEPs to be deleted with 50 and no-limit and it didn't seem to make much difference. This is on a single node cluster, though, and almost all CEPs need to be deleted.

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -197,36 +203,60 @@ func RunK8sCiliumEndpointSyncGC() {
-						if err := ciliumClient.CiliumEndpoints(cep.Namespace).Delete(cep.Name, &meta_v1.DeleteOptions{}); err != nil {
-							scopedLog.WithError(err).Debug("Unable to delete CEP")
+			perCEPFetch:
+				for time.Now().Before(loopStop) { // stop the loop if it takes too long
```
What can take too long?

tgraf replied:
It makes sure that a single controller run does not exceed the interval period

raybejjani replied:
This is really guarding against an infinite loop. I picked the interval period since it was a well-defined time. I'm mostly worried about not calling `break` somewhere in the code.

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -197,36 +203,60 @@ func RunK8sCiliumEndpointSyncGC() {
+					}
+					if ceps.Continue != "" {
+						// there is more data, continue
+						continue perCEPFetch
```
Add a relaxing sleep here to pace the garbage collector.

raybejjani replied:
I added it to the start of the loop since there are two continue cases we encounter.

raybejjani reviewed (commented):
I'll run some more benchmarks but I updated the code a little.

raybejjani committed:
k8s: CEP GC runs in steps

On large clusters the CEP GC would fetch the entire list of CEPs. These
can grow to significant sizes (100kB in one instance) and this causes
extreme degradation in kube-apiserver. We now iterate 10 CEPs at a
time, thus lowering the load we put on the cluster.

Signed-off-by: Ray Bejjani <ray@covalent.io>
